src/linker/Linker.Dataflow/DataFlowUtilities.cs

```diff
@@ -0,0 +1,108 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Mono.Cecil;
+
+namespace Mono.Linker.Dataflow
+{
+	static class DataFlowUtilities
+	{
+		internal static string GetDynamicallyAccessedMemberTypesDescription (DynamicallyAccessedMemberTypes memberTypes)
+		{
+			if (memberTypes == DynamicallyAccessedMemberTypes.All)
+				return DynamicallyAccessedMemberTypes.All.ToString ();
+
+			var results = new List<DynamicallyAccessedMemberTypes> ();
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicConstructors))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicConstructors);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicConstructors))
+				results.Add (DynamicallyAccessedMemberTypes.PublicConstructors);
+			else if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.DefaultConstructor))
+				results.Add (DynamicallyAccessedMemberTypes.DefaultConstructor);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicMethods))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicMethods);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicMethods))
+				results.Add (DynamicallyAccessedMemberTypes.PublicMethods);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicProperties))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicProperties);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicProperties))
+				results.Add (DynamicallyAccessedMemberTypes.PublicProperties);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicFields))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicFields);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicFields))
+				results.Add (DynamicallyAccessedMemberTypes.PublicFields);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicEvents))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicEvents);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicEvents))
+				results.Add (DynamicallyAccessedMemberTypes.PublicEvents);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.NonPublicNestedTypes))
+				results.Add (DynamicallyAccessedMemberTypes.NonPublicNestedTypes);
+
+			if (memberTypes.HasFlag (DynamicallyAccessedMemberTypes.PublicNestedTypes))
+				results.Add (DynamicallyAccessedMemberTypes.PublicNestedTypes);
+
+			if (results.Count == 0)
+				return DynamicallyAccessedMemberTypes.None.ToString ();
+
+			return string.Join (" | ", results.Select (r => r.ToString ()));
```

Comment:
Do we actually want such output? It can produce almost unreadable text

Reply:
It should recreate the text used in the attribute in source code pretty well. By far the most common case is annotations with one or two flags from the enum, so that should be fine.

I must admit I don't understand the scenario when this would produce a mess. The only potential is "I want to include everything", but for that we have the `All` value which is handled explicitly.

Reply:
We used this to dump to user two sets of enums and let him compare. This is very hard as most of the options are very similar. I don't know if the common path will use 2, 3 or more values but I think it'd be better to re-word the error message to allow us to report only single enum value instead.

(ignoring the aspect that this is C#-ish in IL-based tool)

Reply:
I'd like to address this in a separate PR as it's mostly unrelated to this one (the validation introduced in this PR doesn't actually use this method, I just moved it out of the already large `ReflectionMethodBodyScanner`).

I added note about this to https://github.com/mono/linker/issues/1157 which should address the warning message where this is used from.

Reply:
ok